### ISSUE TITLE
Potential inconsistency with usage of `internald` in the RemoteValue serialization

### ISSUE THREAD

**Issue by lutien** [w3c/webdriver-bidi#285]
In the [description of RemoteValue section](https://w3c.github.io/webdriver-bidi/#data-types-protocolValue-RemoteValue) it says:

> In case of duplicated objects in the same RemoteValue, the value is provided only for one of the remote values, while the unique-per-ECMAScript-object internalId is provided for all the duplicated objects for a given serialization.

Then all the type definitions expect of primitives contain `internald`, at the same time when we look at [serialization section](https://w3c.github.io/webdriver-bidi/#data-types-protocolValue-RemoteValue-serialization), we have `set internal ids if needed` point only for array, map, object, set, node. Judging by the description it looks like it should be applied for every RemoteValue type.

**Comment by jgraham** [comment]
I think the fundamental issue here is really this:

In the BiDi serialization spec right now, something like `Date` doesn't serialize it's properties, and so can't participate in a cycle. Therefore _per the current spec_ we don't need to call `set internal ids if needed` and should drop the `internalId` property.

**But**, that means we can create JS objects that can't be correctly serialized e.g. if we do:
```
a = new Date();
a.foo = "bar"
```
We don't serialize the `foo` property, so we can't recreate the same value later. That also applies to cycles (e.g. setting `a.foo = a`), but is not unique to them. That makes some sense since we also currently can't deserialize additional properties on these objects.

So I think the question here is really: do we want to support additional properties on all objects? If we do then we need to change the serialization to something like:
```
DateLocalValue = {
  type: "date",
  value: {
    value: text
    properties: MappingLocalValue
  }
}
```
and also update the spec algorithms to (de)serialize properties not on the built in prototype object as we do for generic objects.

**Comment by lutien** [comment]
I guess it makes sense. @whimboo and @juliandescottes, what do you think?
@jgraham, and what about the objects which have only `handle` and no value (like `WeakMap`), do then they need `internalId`?

**Comment by juliandescottes** [comment]
> do we want to support additional properties on all objects?

I guess we should resolve this first. From a DevTools perspective, we would need a way to access all properties from a non-primitive object, so I think it would make sense to support extra properties. Eg if you want to inspect an object in the console ...

Assuming we go for that and serialize extra properties, would we consistently set internal ids on all non-primitive objects, or only if we detect that there are extra properties?

> what about the objects which have only handle and no value (like WeakMap), do then they need internalId?

I think we can apply the same argument as for Date etc... They can't participate in a cycle so we don't need to set internalIds on them. But if we support serializing extra properties, then they might participate in a cycle and we need internalId as well?

I don't know if there's any added value in making the spec consistent before we resolve the question around serializing extra properties? We could quickly add an `issue` below https://w3c.github.io/webdriver-bidi/#data-types-protocolValue-RemoteValue to highlight that the text is inconsistent with the rest of the spec. (and as far as the implementation goes, I think it's fine to assume that we only set internal ids on values which can participate in a cycle?)

**Comment by juliandescottes** [comment]
Adding needs-discussion to resolve the topic around extra properties serialization

**Comment by jgraham** [comment]
> From a DevTools perspective, we would need a way to access all properties from a non-primitive object

This is still possible of course, but requires an extra roundtrip using the `handle` to turn the internal representation into something we can serialize e.g. `script.callFunction(obj => Object.getOwnPropertyDescriptors(obj), [obj])`. So there's a little bit of a tradeoff here. I think the original thinking was that we should serialize the object as a copy where that's most useful, and the handles are more of a convenience. But the fact that we don't serialize the complete object graph for certain objects does seem surprising.

**Comment by juliandescottes** [comment]
Yes we can workaround with `callFunction`, but that means an extra round trip for each and every non-primitive, even if there are no extra properties.

**Comment by juliandescottes** [comment]
Discussed with James, and even if we had the object's extra properties, for a proper devtools support we would still miss prototype properties, so it's better to focus on whatever makes sense for automation here since the devtools scenario will require an extra round trip anyway (potentially could have a dedicated command rather than relying on callFunction, but that's another topic)

**Comment by jgraham** [comment]
I still wonder if we should guarantee that we serialize all the own properties of an object, so that (given infinite serilization depth) you always get the full graph of direct references (excluding prototypes).

The main problem I see is that for plain `Object` we currently put the properties into the `value` field, whereas for values with a direct serialization of their internal state (e.g. `Date`) we put the serialized state in the `value` field and there's nowhere for additional properties. So we could add `properties` as a field to cover all cases, and avoid giving `Object` a value, (but then `Array` would be weird), or we could make a more complex `value` field for the copyable types with fields for both the internal value and the additional properties.
